apis/apim/v1alpha1/common.go

```diff
@@ -0,0 +1,50 @@
+package v1alpha1
+
+import (
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	"github.com/kuadrant/kuadrant-controller/pkg/common"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Interface to be implemented by Policy's status struct
+//+kubebuilder:object:generate=false
+type PolicyStatus interface {
+	GetObservedGeneration() int64
+	GetConditions() []metav1.Condition
+}
+
+func (r RateLimitPolicyStatus) GetObservedGeneration() int64 {
```

Comment:
this method in its own ratelimitpolicy_status.go file?

Reply:
Types are in `v1alpha1` package and implementing the interface in a different package is not allowed in Go but if you mean to create a new file called `ratelimitpolicy_status.go` then I'd say we can do that once `common.go` is large enough to break into different files.

Reply:
Sorry, I was referring to the `ratelimitpolicy_types.go`.  So the implementation of the interface happens in its own file for rate lmit types and auth types.

>  implementing the interface in a different package is not allowed in Go 

What? where is that statement come from? I am missing something?

`myinterface/myinterface.go` file
```go
package myinterface

type PolicyStatus interface {
	GetObservedGeneration() int64
	GetConditions() []int64
}
```

`myinterfaceimpl/myinterfaceimpl.go` file

```go
package myinterfaceimpl

type PolicyStatusImpl int64

func (s PolicyStatusImpl) GetObservedGeneration() int64 {
	return int64(s)
}

func (s PolicyStatusImpl) GetConditions() []int64 {
	return []int64{int64(s)}
}
```

`main.go` file

```go
package main

import (
	"fmt"
	"my-module/myinterface"
	"my-module/myinterfaceimpl"
)

func main() {
	var xs myinterface.PolicyStatus = myinterfaceimpl.PolicyStatusImpl(5)
	fmt.Println(xs)
}
```

```
❯ go run main.go 
5
```

Reply:
I was previously trying to implement the interface in the `controllers/apim` package for status structs that are in `v1alpha` package.
```
invalid receiver type "github.com/kuadrant/kuadrant-controller/apis/apim/v1alpha1".RateLimitPolicyStatus (type not defined in this package)
```

Reply:
What you cannot do is to implement a method for type A in a package where A is not defined. If your type is defined in some package P, the type methods can be defined in multiple files, but all belonging to package P